### DISCUSSION TITLE
cFE 6.7, CMake and target-specific variables

### DISCUSSION THREAD

**Discussion by clemejm1-apl** [nasa/cFS#558] [Q&A] [answered]
I'm updating a legacy codebase to cFS 6.7 and CMake. Where are target-specific cmake variables meant to be set?  We have multiple targets defined and its more or less working, but we have some apps and libs that set preprocessor `#defines` or conditionally compile different files based on whether they're built for pc-linux or rtems. 

One way is to set them in  `toolchain-$TGT.cmake` which works but it's not the type of thing I'd expect to find in a toolchain file. 

Perhaps there is a cmake hook I don't know about? Maybe something like what's discussed in https://github.com/nasa/cFS/discussions/403 ?

Alternatively, how is the current target exposed to app-level CMakeLists.txt? 

I'm familiar with CMake but still learning cFS so my apologies if I'm missing something obvious.

**Comment by skliper** [comment, marked as answer]
If it's platform dependent maybe use `CFE_SYSTEM_PSPNAME`?  Note you can get all the defined variables from the current context via the method described here:
https://stackoverflow.com/questions/9298278/cmake-print-out-all-accessible-variables-in-a-script

**Reply by clemejm1-apl** [reply]
Thanks. Digging through the defined variables, it appears `${TGTNAME}` is exposed at the app layer so you can key off that. That gets me past my immediate roadblock. 

I'd still like to know if there's a preferred way to export mission-wide custom target-specific CMake variables.

**Reply by skliper** [reply]
If I'm understanding correctly, you could transition to a cmake config approach for your app(s) like *_osconfig.cmake which can then be target specific (this is likely where we are going with cFE and the rest eventually for cmake source selection to be more seamless) and then pull in the right config in arch_build_custom.cmake (similar to what's suggested in #403).  Or add the variables to the arch_build_custom.cmake w conditionals on ${TGTNAME}.  There's also the toolchain approach you already mentioned but those are often more platform centric vs target.
